lib/consumer.js

```diff
@@ -5,82 +7,77 @@ var amqp = require('amqplib'),
   utils = require('./helpers/utils'),
   parsers = require('./helpers/message-parsers');
 
-var amqpConnection, amqpChannel, amqpUrl, amqpIntervalId, amqpConfig, amqpState;
-var amqpConnPool = [];
-var amqpConsumptions = [];
-var amqpReconnection = false;
-var amqpStates = {
-  DISCONNECTED: -1,
-  CONNECTING: 0,
-  CONNECTED: 1
-};
-
-amqpState = amqpStates.DISCONNECTED;
-
-var connect = function (url) {
-  return new Promise(function (_resolve) {
-    if (amqpState === amqpStates.DISCONNECTED) {
-      amqpState = amqpStates.CONNECTING;
+var amqpConnection, amqpChannel, amqpConfig;
+var amqpReconnect = false;
+var amqpQueues = [];
 
-      amqpUrl = utils.getValidUrl([amqpUrl, url, amqpConfig.amqpUrl, process.env.AMQP_URL, 'amqp://localhost']);
+var connect = () => {
+  if (!amqpConnection) {
+    amqpChannel = null;
+    amqpConnection = new Promise((resolve, reject) => {
+      let amqpUrl = utils.getValidUrl([amqpConfig.amqpUrl, process.env.AMQP_URL, 'amqp://localhost']);
 
       amqp.connect(amqpUrl, { clientProperties: { hostname: process.env.HOSTNAME } })
-      .then(function (_connection) {
-        amqpConnection = _connection;
-        Logger.info('[BMQ-CONSUMER] Is trying to connect');
+      .then((_connection) => {
 
-        amqpIntervalId = clearInterval(amqpIntervalId);
+        _connection.on('close', (err) => {
+          Logger.error(err);
+          amqpConnection = null;
+          amqpChannel = null;
+        });
 
-        amqpConnection.on('close', reconnect);
-        amqpConnection.on('error', reconnect);
+        _connection.on('error', Logger.error);
 
-        return amqpConnection.createChannel();
+        amqpConnection = _connection;
+        resolve(amqpConnection);
       })
-      .then(function (_channel) {
-        amqpChannel = _channel;
-        amqpChannel.prefetch(amqpConfig.amqpPrefetch);
+      .catch((err) => {
+        amqpConnection = null;
+        reject(err);
+      });
+    });
+  }
+
+  return amqpConnection;
+};
 
-        Logger.info('[BMQ-CONSUMER] Is now connected and ready to consume messages');
+var createChannel = () => {
+  if (!amqpChannel) {
+    amqpChannel = new Promise((resolve, reject) => {
+      amqpConnection.createChannel()
+        .then((_channel) => {
+          _channel.prefetch(amqpConfig.amqpPrefetch);
 
-        amqpState = amqpStates.CONNECTED;
+          Logger.info('[BMQ-CONSUMER] Is now connected and ready to consume messages');
 
-        amqpIntervalId = clearInterval(amqpIntervalId);
+          _channel.on('close', (err) => {
+            if (err) Logger.error('ERROR:', err);
+            amqpChannel = null;
+          });
 
-        amqpChannel.on('close', reconnect);
-        amqpChannel.on('error', reconnect);
+          _channel.on('error', Logger.error);
 
-        while (amqpConnPool.length > 0) {
-          amqpConnPool.shift()(amqpChannel);
-        }
+          amqpChannel = _channel;
+          resolve(amqpChannel);
```

Comment:
It is our intent because the function should return the ``amqpChannel``, and that is done below (L77)

The resolve here is just for producers that ocured while disconnected, so they can all resolve because ``amqpChannel`` in these cases is a promise, and is returned too on L77.

It should be commented